### DISCUSSION TITLE
How to create a TeXstudio macro to remove extra brackets?

### DISCUSSION THREAD

**Discussion by fernandozucatelli** [texstudio-org/texstudio#1884] [Q&A] [answered]
This question is the same I made at the TeXStackExchange Forum in:
https://tex.stackexchange.com/questions/619761/how-to-create-a-texstudio-macro-to-remove-extra-brackets.

I have a software to write mathematical expressions that let me quite faster type while teaching. This software have a LaTeX translation tool that works fine, but it adds as many brackets as it wants.

I have been using the replace tool of TeXstudio with regular expressions (`regexp`) for a while in order to correct some issues and recently I found how to remove extra brackets in expressions. So I can transform `{{any}}` into `{any}`.

I would like to have a TeXstudio macro to perform all replacements, but I got some difficult 
writing the correct code for the macro version of the find and replace syntax.

The following links helped me to create the macro as far as I could.

 1. https://tex.stackexchange.com/questions/432218/how-to-insert-submatches-from-regular-expression-in-texstudio
 2. https://tex.stackexchange.com/questions/272690/broken-regexp-replace-in-texstudio-macros
 3. https://tex.stackexchange.com/questions/510503/texstudio-how-to-search-regexp-to-the-next-bracket
 4. https://www.w3schools.com/jsref/jsref_obj_regexp.asp
 5. https://htmlpreview.github.io/?https://github.com/texstudio-org/texstudio/master/utilities/manual/usermanual_en.html#SECTION33

The following figure shows the Find/Replace syntax that works and the macro version I tried without success.

[![enter image description here][1]][1]

The Find syntax `{{([^\{\}]+)}}` also works with `{{([^\{\}]*)}}`.

I don't know what I should do with the "`\1`" part I used in the Replace when it goes into the macro.

The operation performed by Find and Replace works as expected.

[![enter image description here][2]][2]

But when applying the macro I get `null` as result. I clearly don't know the correct syntax to do the substitutions properly.

[![enter image description here][3]][3]

A MWE follows, this MWE don't need to be executed.

    \begin{equation}
      {{1}} \frac{{a}}{{b}} \dfrac{{num}}{{den}}
      {{a + b}}^{{c + d}}
      {{m + n}
      }
    \end{equation}

The SCRIPT I'm trying to create is:

    %SCRIPT
    options = "g"
    scope = editor.document().cursor(0, 0, -1)
    editor.replace(/\{\{[^\{\}]*\}\}/g,
    	options,scope,
    	/\{\\1\}/)

I also tried using a fixed string with `"{\\1}"`

    editor.replace(/\{\{[^\{\}]*\}\}/g,
    	options,scope,
    	"{\\1}")

and also some variations like `/\{$1\}/` without success.

----------
**Edit**

The regexp in Find syntax `/\{\{[^\{\}]+\}\}/` (or `*` instead of `+`) seems to work properly, it finds every double brackets. Using a replace syntax with a fixed string such as `"{A}"` does the correct replacement as presented at the following figure.

[![enter image description here][4]][4]

  [1]: https://i.stack.imgur.com/lZtIg.png
  [2]: https://i.stack.imgur.com/1sBDQ.png
  [3]: https://i.stack.imgur.com/hVlpA.png
  [4]: https://i.stack.imgur.com/OXXy6.png

**Comment by sunderme** [comment, marked as answer]
```
%SCRIPT
editor.replace(/{{(.*)}}/,"g","{\\1}")
```
seems to work.

scope is optional and can be left out if it is all.
The cursor needs to be set differently in txs4.x , see manual.

... and you javascript regex misses the brackets `()`.

**Reply by fernandozucatelli** [reply]
Thanks.
Scope and the missing brackets were the keys.
On the other hand, the search syntax with `(.*)` misses some brackets on my test subject, therefore I kept `/{{([^\{\}]*)}}/`.

```
%SCRIPT
editor.replace(/{{([^\{\}]*)}}/, "g", "{\\1}")
```
